homeassistant/components/recorder/__init__.py

```diff
@@ -422,93 +271,25 @@ def _close_connection(self):
                 run.closed_incorrect = True
                 run.end = self.recording_start
                 _LOGGER.warning("Ended unfinished session (id=%s from %s)",
                                 run.run_id, run.start)
                 session.add(run)
 
-                _LOGGER.warning("Found unfinished sessions")
-
-            self._run = recorder_runs(
+            self.run_info = recorder_runs(
                 start=self.recording_start,
                 created=dt_util.utcnow()
             )
-            self._commit(session, self._run)
+            session.add(self.run_info)
 
     def _close_run(self):
         """Save end time for current run."""
-        with session_scope() as session:
-            self._run.end = dt_util.utcnow()
-            self._commit(session, self._run)
-        self._run = None
-
-    def _purge_old_data(self, _=None):
-        """Purge events and states older than purge_days ago."""
-        from homeassistant.components.recorder.models import Events, States
-
-        if not self.purge_days or self.purge_days < 1:
-            _LOGGER.debug("purge_days set to %s, will not purge any old data.",
-                          self.purge_days)
-            return
-
-        purge_before = dt_util.utcnow() - timedelta(days=self.purge_days)
-
-        def _purge_states(session):
-            deleted_rows = session.query(States) \
-                                  .filter((States.created < purge_before)) \
-                                  .delete(synchronize_session=False)
-            _LOGGER.debug("Deleted %s states", deleted_rows)
-
-        with session_scope() as session:
-            if self._commit(session, _purge_states):
-                _LOGGER.info("Purged states created before %s", purge_before)
-
-        def _purge_events(session):
-            deleted_rows = session.query(Events) \
-                                  .filter((Events.created < purge_before)) \
-                                  .delete(synchronize_session=False)
-            _LOGGER.debug("Deleted %s events", deleted_rows)
-
-        with session_scope() as session:
-            if self._commit(session, _purge_events):
-                _LOGGER.info("Purged events created before %s", purge_before)
-
-        # Execute sqlite vacuum command to free up space on disk
-        if self.engine.driver == 'sqlite':
-            _LOGGER.info("Vacuuming SQLite to free space")
-            self.engine.execute("VACUUM")
-
-    @staticmethod
-    def _commit(session, work):
-        """Commit & retry work: Either a model or in a function."""
-        import sqlalchemy.exc
-        for _ in range(0, RETRIES):
-            try:
-                if callable(work):
-                    work(session)
-                else:
-                    session.add(work)
-                session.commit()
-                return True
-            except sqlalchemy.exc.OperationalError as err:
```

Comment:
I find it kinda weird that we do things 3 times and I think that it also added to the confusion of how SQLAlchemy works (instantiating sessions everywhere for example).